tools/clang/test/SemaHLSL/attributes/wavesize_range.hlsl

```diff
@@ -76,7 +74,9 @@ void node07(DispatchNodeInputRecord<INPUT_RECORD> input) { }
 [NodeLaunch("broadcasting")]
 [NumThreads(1,1,1)]
 [NodeMaxDispatchGrid(32,1,1)]
-[WaveSize(4, 8, node07)] /* expected-error{{'WaveSize' attribute requires an integer constant}} */
+// expected-error@+2{{WaveSize arguments must be between 4 and 128 and a power of 2}}
```

Comment:
Conversely, this error seems unhelpful. No one would argue that a function call is between 4 and 128, but it seems an odd thing to assert.

Reply:
The only reason this wasn't emitted before is because it didn't catch the explicit preferred==0 case either, so this was due to missing diagnostic coverage.

With that coverage, it's quite a bit more difficult to avoid emitting this specific diagnostic, without some way of stopping diagnostics as soon as something fails when it's checking for int args (which we currently have no way of knowing outside that function).

In any case, along with the `'WaveSize' attribute requires an integer constant`, this is useful because it more specifically defines the allowed values, isn't it?

Reply:
I don't understand where preferred == 0 comes in. Because the param is invalid, it defaults to 0? Is that it?

Reply:
See where the arg types are checked and the attribute is created here:
https://github.com/microsoft/DirectXShaderCompiler/pull/6172/files#diff-b5ff0eb7c221e9ad11a754e889ea567c35d83ad9f51e89e77ec844ea4fd8eebbL12625

The function `ValidateAttributeIntArg` that checks for int argument type always returns an int.  It just returns `0` if it was invalid.  There's no indication that it emitted an error as far as `ValidateWaveSizeAttributes` is concerned.

The rest of the diagnostic code that looks at this attribute cannot tell the difference between a type error on an argument and an explicit `0` supplied by the user.